examples/step-79/step-79.cc

```diff
@@ -0,0 +1,2755 @@
+                const auto center = face->center();
+                if (std::fabs(center(1) - 0) < 1e-12)
+                  {
+                    face->set_boundary_id(BoundaryIds::no_force);
+                  }
+                else if (std::fabs(center(1) - 1) < 1e-12)
+                  {
+                    if ((std::fabs(center(0) - 3) < .3))
+                      {
+                        face->set_boundary_id(BoundaryIds::down_force);
+                      }
+                    else
+                      {
+                        face->set_boundary_id(BoundaryIds::no_force);
+                      }
+                  }
+                else if (std::fabs(center(0) - 0) < 1e-12)
+                  {
+                    face->set_boundary_id(BoundaryIds::no_force);
+                  }
+                else if (std::fabs(center(0) - 6) < 1e-12)
+                  {
+                    face->set_boundary_id(BoundaryIds::no_force);
+                  }
+              }
+          }
+      }
+  }
+
+
+  // Next, determine the constraints due to boundary values.  The
+  // bottom corners of the domain are kept in place in the $y$
+  // direction -- the bottom left also in the $x$ direction. deal.II
+  // generally thinks of boundary values as attached to pieces of the
+  // boundary, i.e., faces, rather than individual vertices. Indeed,
+  // mathematically speaking, one can not assign boundary values to
+  // individual points for the infinite-dimensional partial
+  // differential equation. But, since we are trying to reproduce a
+  // widely used benchmark, we will do so anyway and keep in mind that
+  // we have a finite-dimensional problem for which imposing boundary
+  // conditions at a single node is valid.
+  template <int dim>
+  void SANDTopOpt<dim>::setup_boundary_values()
+  {
+    boundary_values.clear();
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        for (const auto &face : cell->face_iterators())
+          {
+            if (face->at_boundary())
+              {
+                const auto center = face->center();
+
+                // Check whether the current face is on the bottom
+                // boundary, and if it is whether one of its
+                // vertices might be the bottom left or bottom
+                // right vertex:
+                if (std::fabs(center(1) - 0) < 1e-12)
+                  {
+                    for (unsigned int vertex_number = 0;
+                         vertex_number < GeometryInfo<dim>::vertices_per_cell;
+                         ++vertex_number)
+                      {
+                        const auto vert = cell->vertex(vertex_number);
+
+                        if (std::fabs(vert(0) - 0) < 1e-12 &&
+                            std::fabs(vert(1) - 0) < 1e-12)
+                          {
+                            const unsigned int x_displacement =
+                              cell->vertex_dof_index(vertex_number, 0);
```

Comment:
(here and below) the function returns a value of type ``types::global_dof_index``, please use that instead of ``unsigned int``.